source/extensions/quic_listeners/quiche/envoy_quic_client_connection.cc

```diff
@@ -0,0 +1,144 @@
+
+void EnvoyQuicClientConnection::processPacket(
+    Network::Address::InstanceConstSharedPtr local_address,
+    Network::Address::InstanceConstSharedPtr peer_address, Buffer::InstancePtr buffer,
+    MonotonicTime receive_time) {
+  quic::QuicTime timestamp =
+      quic::QuicTime::Zero() +
+      quic::QuicTime::Delta::FromMicroseconds(
+          std::chrono::duration_cast<std::chrono::microseconds>(receive_time.time_since_epoch())
+              .count());
+  uint64_t num_slice = buffer->getRawSlices(nullptr, 0);
+  ASSERT(num_slice == 1);
+  Buffer::RawSlice slice;
+  buffer->getRawSlices(&slice, 1);
+  quic::QuicReceivedPacket packet(reinterpret_cast<char*>(slice.mem_), slice.len_, timestamp,
+                                  /*owns_buffer=*/false, /*ttl=*/0, /*ttl_valid=*/true,
+                                  /*packet_headers=*/nullptr, /*headers_length=*/0,
+                                  /*owns_header_buffer*/ false);
+  ProcessUdpPacket(envoyAddressInstanceToQuicSocketAddress(local_address),
+                   envoyAddressInstanceToQuicSocketAddress(peer_address), packet);
+}
+
+uint64_t EnvoyQuicClientConnection::maxPacketSize() const {
+  // TODO(danzh) make this variable configurable to support jumbo frames.
+  return Network::MAX_UDP_PACKET_SIZE;
+}
+
+void EnvoyQuicClientConnection::setUpConnectionSocket() {
+  if (connectionSocket()->ioHandle().isOpen()) {
+    file_event_ = dispatcher_.createFileEvent(
+        connectionSocket()->ioHandle().fd(),
+        [this](uint32_t events) -> void { onFileEvent(events); }, Event::FileTriggerType::Edge,
+        Event::FileReadyType::Read | Event::FileReadyType::Write);
+
+    if (!Network::Socket::applyOptions(connectionSocket()->options(), *connectionSocket(),
+                                       envoy::api::v2::core::SocketOption::STATE_LISTENING)) {
+      ENVOY_CONN_LOG(error, "Fail to apply listening options", *this);
+      connectionSocket()->close();
+    }
+  }
+  if (!connectionSocket()->ioHandle().isOpen()) {
+    CloseConnection(quic::QUIC_CONNECTION_CANCELLED, "Fail to setup connection socket.",
+                    quic::ConnectionCloseBehavior::SILENT_CLOSE);
+  }
+}
+
+void EnvoyQuicClientConnection::onFileEvent(uint32_t events) {
+  ENVOY_CONN_LOG(trace, "socket event: {}", *this, events);
+  ASSERT(events & (Event::FileReadyType::Read | Event::FileReadyType::Write));
+
+  if (events & Event::FileReadyType::Write) {
+    OnCanWrite();
+  }
+
+  // It's possible for a write event callback to close the connection, in such case ignore read
+  // event processing.
+  if (connected() && (events & Event::FileReadyType::Read)) {
+    uint32_t old_packets_dropped = packets_dropped_;
+    while (connected()) {
+      // Read till socket is drained.
+      // TODO(danzh): limit read times here.
+      MonotonicTime receive_time = dispatcher_.timeSource().monotonicTime();
+      Api::IoCallUint64Result result = Network::Utility::readFromSocket(
```

Comment:
You should be able to use my new utility to read multiple packets? Can you refactor to use that method and modify if needed?

Reply:
done.